### ISSUE TITLE
Preload JOOQ classes in DSLContext initialization instead of first query execution

### ISSUE THREAD

**Issue by kgribov** [jOOQ/jOOQ#12246]
### Use case:

Hey Lukas!
I and my team are big fans of JOOQ. We used it in our previous company and use it in our new project. Awesome library!
 
We found interesting behaviour in our system - first requests after deployment were always slow (500ms vs 10ms avg). 
In logs I see that JOOQ starts to initialize (prints JOOQ welcome message and etc) only on first SQL query. I wrote some small test to profile it and check what is happening in JOOQ and why it is so slow.

![image](https://user-images.githubusercontent.com/2950277/126625943-97eef067-d7e7-4735-b66b-ba52e451d574.png)
![image](https://user-images.githubusercontent.com/2950277/126625998-c8e74771-21ba-4aff-bf2d-aba39734ad3d.png)

After JOOQ loads all this classes (like `org.jooq.impl.Convert`) - all queries starts to work fast. 
Example of code we use for query execution:
```
val query = dslContext.insertQuery(Tables.TEST).apply { setRecord(TestRecord(randomUUID(), Instant.now())) }
transactionManager.inTransaction(ConnectionMode.REQUIRE_NEW) { connection ->
      DSL.using(connection, dslContext.settings())
          .fetch(dslContext.render(query), *params).into(TestRecord::class.java)
  }
```

### Possible solution you'd like to see:

To have some setting's option or method to preload all major JOOQ classes and settings, so it doesn't affect query execution. Or may be there is some practise how we can do this preload.

### Possible workarounds:

As a workaround/hack we can have "dump" table and insert some fake records before we start to execute production queries. 

### Versions:

- jOOQ: 3.15.0
- Java: Kotlin 1.5.20 JDK 16.0.1
- Database (include vendor): Postgresql 
- OS: Mac OS
- JDBC Driver (include name if inofficial driver):

**Comment by lukaseder** [comment]
Thanks for your message. The work of loading classes and the meta data has to be done. You can easily time this by triggering such load work at your earliest convenience. It's much harder for jOOQ to control, and I don't think we should overengineer any delaying mechanism here.

However, do check a few things:

1. Are you sure it's jOOQ working, and not some third party? I remember quite a few cases where Spring instrumented a lot of jOOQ classes
2. What *exactly* is it you're loading? Can you dig a bit further in the profiler (try increasing the sample rate, or even switch to profiling rather than sampling)? Maybe there is something that could be avoided - hard to say from the high level view of your sampling output.

**Comment by kgribov** [comment]
thanks a lot for your response!

1. We don't have Spring or other big frameworks. We have very small abstraction on top of JOOQ. But I will double check
2. ` Can you dig a bit further in the profiler` Will do ASAP!
more detailed sampling on first screen is - 
<img width="1401" alt="image" src="https://user-images.githubusercontent.com/2950277/126681554-af2691ae-7895-47ba-8111-706bce899ce9.png">

If I m not wrong - loading of Jackson classes takes most of time

**Comment by lukaseder** [comment]
1. Can you try to change the sample size? 100ms or 10/s is not too fine grained
2. Can you benchmark this code in throwaway classloaders repeatedly, to make sure no other side effects are in play?
3. Do you actually use Jackson or is it on the classpath through some transitive dependency?

In any case, I think we can delay the work until it is actually needed, which is never in a lot of applications

**Comment by lukaseder** [comment]
Huh, looking at this again, you're requesting the work be done *earlier* rather than *later* (or not at all). That doesn't seem reasonable in my opinion. Lazy classloading is a very good feature in Java. It helps prevent unnecessary work, e.g. when a feature isn't actually needed. In this case, some users might have Jackson on their classpaths, but they never need the feature from jOOQ, so they shouldn't have to run this code.

But making sure a library is ready for usage when the first user runs a query is a different story. jOOQ shouldn't eagerly load all of its code, including a possibly huge meta model. That would be the wrong solution for a large amount of users. If you need that to be done, it's easy to run a quick query on startup, long before a user accesses the system.

**Comment by kgribov** [comment]
hey! sorry for late response.
`Can you try to change the sample size?` yep, I did some runs with Yourkit profiler, where default sample size is 20ms. I did both sampling and tracing - 
![image](https://user-images.githubusercontent.com/2950277/126979841-fab4a685-c7fc-4e91-8911-8684e80ed77b.png)
![image](https://user-images.githubusercontent.com/2950277/126979885-ca827068-df1e-486b-a34d-fe6bcaf5d347.png)

Even if you don't have JSON/JSONB in your code - jOOQ will try to load Jackson anyway, because it is a part of `org.jooq.impl.Conver` static context. F.e. if we want to get UUID value, Convert class will be involved -
![image](https://user-images.githubusercontent.com/2950277/126980822-9509a22a-7a32-4c29-af1f-45f32fede77e.png)

`Lazy classloading is a very good feature in Java` - totally agree! On the other hand, sometimes you want to be sure, that your code is ready to do some work.

`In this case, some users might have Jackson on their classpaths, but they never need the feature from jOOQ` - unfortunately it is wrong, if jOOQ uses  `DefaultBinding` class.

I made a small hack and added just this line to my code - `Class.forName("org.jooq.impl.Convert")`. First query duration reduced from ~500ms to ~200ms.

200ms is still not so optimal as it could be (10-20ms regular queries), in profiler I see lot of other class loading stuff - lot of micro loads, but in result it takes 200ms haha

So as a quick solution we can do this hack with `Class.forName` or introduce some test-query (better solution), which will touch this Convert class and other classes we use for our queries.

I think you can close this issue, but I think it will be better to move Jackson load from `org.jooq.impl.Convert` static context away. If you have any guidelines how to properly "warm up" jOOQ and queries - will be happy to read it.

Thanks a lot for your time!

**Comment by lukaseder** [comment]
> unfortunately it is wrong, if jOOQ uses `DefaultBinding` class.

My point is that if you don't have Jackson on the classpath, then that static initialiser code will be aborted with a `ClassNotFoundException` and jOOQ will never use Jackson - it should be much faster.

> but I think it will be better to move Jackson load from `org.jooq.impl.Convert` static context away.

Yes, that's what I meant by *In any case, I think we can delay the work until it is actually needed, which is never in a lot of applications*

It can be loaded the first time a JSON data structure is converted to a POJO (possibly never), but that work will have to be done at some point if the feature is being used.

> If you have any guidelines how to properly "warm up" jOOQ and queries - will be happy to read it.

I don't think it would be reasonable to document this, because initialisation happens when it needs to happen. In your case, why not just run something like:

```java
ctx.selectCount()
    .from(MY_TABLE) // Force initialisation of generated code
    .where(falseCondition())
    .fetchOneInto(int.class); // Force initialisation of this Convert class
```

But again, don't consider this to be documentation. Everything internal can change at any time. If you want to remove startup times entirely, use AOP (e.g. via GraalVM, see also https://github.com/quarkiverse/quarkus-jooq).

**Comment by lukaseder** [comment]
I've added a task to https://github.com/jOOQ/jOOQ/issues/11894, which deals with refactoring of the conversion logic currently implemented in `Convert`. With the refactored logic, it will be much easier to delay initialisation of third party utilities to the moment when they are actually needed, which will help most users who do not use these features.

I'm closing this as "won't fix", because preloading is something that I don't think can be implemented easily.

**Comment by lukaseder** [comment]
Task about delaying initialisation is here: https://github.com/jOOQ/jOOQ/issues/12262
